cpp/src/arrow/util/rle_encoding_internal.h

```diff
@@ -84,32 +83,279 @@ namespace util {
 /// (total 26 bytes, 1 byte overhead)
 //
 
-/// Decoder class for RLE encoded data.
-class RleDecoder {
+template <typename T>
+class RleRunDecoder;
+
+class RleRun {
  public:
-  /// Create a decoder object. buffer/buffer_len is the decoded data.
-  /// bit_width is the width of each value (before encoding).
-  RleDecoder(const uint8_t* buffer, int buffer_len, int bit_width)
-      : bit_reader_(buffer, buffer_len),
-        bit_width_(bit_width),
-        current_value_(0),
-        repeat_count_(0),
-        literal_count_(0) {
-    ARROW_DCHECK_GE(bit_width_, 0);
-    ARROW_DCHECK_LE(bit_width_, 64);
-  }
+  /// Enough space to store a 64bit value
+  using raw_data_storage = std::array<uint8_t, 8>;
+  using raw_data_const_pointer = const uint8_t*;
+  using raw_data_size_type = int32_t;
+  /// The type of the size of either run, between 1 and 2^31-1 as per Parquet spec
+  using values_count_type = int32_t;
+  /// The type to represent a size in bits
+  using bit_size_type = int32_t;
+
+  /// The decoder class used to decode a single run in the given type.
+  template <typename T>
+  using DecoderType = RleRunDecoder<T>;
 
-  RleDecoder() : bit_width_(-1) {}
+  constexpr RleRun() noexcept = default;
+  constexpr RleRun(const RleRun&) noexcept = default;
+  constexpr RleRun(RleRun&&) noexcept = default;
 
-  void Reset(const uint8_t* buffer, int buffer_len, int bit_width) {
-    ARROW_DCHECK_GE(bit_width, 0);
-    ARROW_DCHECK_LE(bit_width, 64);
-    bit_reader_.Reset(buffer, buffer_len);
-    bit_width_ = bit_width;
-    current_value_ = 0;
-    repeat_count_ = 0;
-    literal_count_ = 0;
-  }
+  explicit RleRun(raw_data_const_pointer data, values_count_type values_count,
+                  bit_size_type value_bit_width) noexcept;
+
+  constexpr RleRun& operator=(const RleRun&) noexcept = default;
+  constexpr RleRun& operator=(RleRun&&) noexcept = default;
+
+  /// The number of repeated values in this run.
+  [[nodiscard]] constexpr values_count_type ValuesCount() const noexcept;
```

Comment:
Again, `[[nodiscard]]` does not seem useful on such simple accessor methods. Nothing dangerous happens if you ignore the result...